core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java

```diff
@@ -168,68 +170,70 @@ protected PlanNodeStatsEstimate visitNotExpression(NotExpression node, Void cont
+                    return estimateLogicalAnd(node.getTerms());
                 case OR:
-                    return estimateLogicalOr(node.getLeft(), node.getRight());
+                    return estimateLogicalOr(node.getTerms());
             }
             throw new IllegalArgumentException("Unexpected binary operator: " + node.getOperator());
         }
 
-        private PlanNodeStatsEstimate estimateLogicalAnd(Expression left, Expression right)
+        private PlanNodeStatsEstimate estimateLogicalAnd(List<Expression> terms)
         {
             // first try to estimate in the fair way
-            PlanNodeStatsEstimate leftEstimate = process(left);
-            if (!leftEstimate.isOutputRowCountUnknown()) {
-                PlanNodeStatsEstimate logicalAndEstimate = new FilterExpressionStatsCalculatingVisitor(leftEstimate, session, types).process(right);
-                if (!logicalAndEstimate.isOutputRowCountUnknown()) {
-                    return logicalAndEstimate;
+            PlanNodeStatsEstimate estimate = process(terms.get(0));
+            for (int i = 1; i < terms.size(); i++) {
+                estimate = new FilterExpressionStatsCalculatingVisitor(estimate, session, types).process(terms.get(i));
+
+                if (estimate.isOutputRowCountUnknown()) {
+                    break;
                 }
             }
 
+            if (!estimate.isOutputRowCountUnknown()) {
+                return estimate;
+            }
+
             // If some of the filters cannot be estimated, take the smallest estimate.
             // Apply 0.9 filter factor as "unknown filter" factor.
-            PlanNodeStatsEstimate rightEstimate = process(right);
-            PlanNodeStatsEstimate smallestKnownEstimate;
-            if (leftEstimate.isOutputRowCountUnknown()) {
-                smallestKnownEstimate = rightEstimate;
-            }
-            else if (rightEstimate.isOutputRowCountUnknown()) {
-                smallestKnownEstimate = leftEstimate;
-            }
-            else {
-                smallestKnownEstimate = leftEstimate.getOutputRowCount() <= rightEstimate.getOutputRowCount() ? leftEstimate : rightEstimate;
-            }
-            if (smallestKnownEstimate.isOutputRowCountUnknown()) {
+            Optional<PlanNodeStatsEstimate> smallest = terms.stream()
+                    .map(this::process)
+                    .filter(termEstimate -> !termEstimate.isOutputRowCountUnknown())
+                    .sorted(Comparator.comparingDouble(PlanNodeStatsEstimate::getOutputRowCount))
+                    .findFirst();
+
+            if (smallest.isEmpty()) {
                 return PlanNodeStatsEstimate.unknown();
             }
-            return smallestKnownEstimate.mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
+
+            return smallest.get().mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
```

Comment:
It's hard to replicate the old behavior because it relies on recursive calls to this method. There's no science to this, though (the choice of multiplying by 0.9 is completely arbitrary) and the old behavior was a little flawed anyway. For instance, these two expressions might return different stats: `(a AND b) AND c` vs `(a AND c) AND b` if computing the stats for `b` returned "unknown".